File: src/main/java/cc/mirukuneko/realtrainmodrenewed/entity/TrainSeatEntity.java
package cc.mirukuneko.realtrainmodrenewed.entity;

import net.minecraft.network.protocol.Packet;
import net.minecraft.network.protocol.game.ClientGamePacketListener;
import net.minecraft.network.protocol.game.ClientboundAddEntityPacket;
import net.minecraft.network.syncher.EntityDataAccessor;
import net.minecraft.network.syncher.EntityDataSerializers;
import net.minecraft.network.syncher.SynchedEntityData;
import net.minecraft.server.level.ServerEntity;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.damagesource.DamageSource;
import net.minecraft.world.entity.*;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.storage.ValueInput;
import net.minecraft.world.level.storage.ValueOutput;
import net.minecraft.world.phys.Vec3;

public class TrainSeatEntity extends Entity {
    private static final float HITBOX_WIDTH = 0.9f;
    private static final float HITBOX_HEIGHT = 0.25f;
    private static final EntityDataAccessor<Integer> SEAT_INDEX = SynchedEntityData.defineId(TrainSeatEntity.class, EntityDataSerializers.INT);

    public TrainEntity train;
    private int cachedTrainId = -1;

    public TrainSeatEntity(EntityType<? extends TrainSeatEntity> type, Level level) {
        super(type, level);
    }

    @Override
    protected void defineSynchedData(SynchedEntityData.Builder builder) {
        builder.define(SEAT_INDEX, -1);
    }

    public void setTrain(TrainEntity train, int seatIndex) {
        this.train = train;
        entityData.set(SEAT_INDEX, seatIndex);
    }

    public TrainEntity getTrain() {
        if (train != null) return train;
        if (level() == null) return null;
        Entity vehicle = getVehicle();
        if (vehicle instanceof TrainEntity t) {
            train = t;
            return t;
        }
        return null;
    }

    public int getSeatIndex() { return entityData.get(SEAT_INDEX); }

    public double getPassengersRidingOffset() {
        return 0.7;
    }

    @Override
    public InteractionResult interact(Player player, InteractionHand hand, Vec3 location) {
        if (player.isCrouching()) return InteractionResult.PASS;
        if (!level().isClientSide()) {
            player.startRiding(this, true, false);
            return InteractionResult.CONSUME;
        }
        return InteractionResult.SUCCESS;
    }

    @Override
    public void tick() {
        super.tick();
        if (level().isClientSide()) {
            TrainEntity t = getTrain();
            if (t != null) {
                // Position follows train
            }
        }
    }

    @Override
    public void readAdditionalSaveData(ValueInput tag) {}
    @Override
    public void addAdditionalSaveData(ValueOutput tag) {}

    @Override
    public Packet<ClientGamePacketListener> getAddEntityPacket(ServerEntity entity) {
        return new ClientboundAddEntityPacket(this, entity);
    }

    @Override
    public EntityDimensions getDimensions(Pose pose) {
        return EntityDimensions.scalable(HITBOX_WIDTH, HITBOX_HEIGHT);
    }

    @Override
    public boolean isPushable() { return false; }
    @Override
    public boolean isPickable() { return true; }
    public boolean canBeCollidedWith() { return false; }

    public void attachToTrain(TrainEntity train, int seatIndex) { setTrain(train, seatIndex); }
    public boolean belongsToTrain(int trainId) { return train != null && train.getId() == trainId; }

    @Override
    protected void addPassenger(Entity passenger) {
        super.addPassenger(passenger);
        if (passenger instanceof Player) {
            passenger.setYRot(getYRot());
        }
    }

    @Override
    public boolean hurtServer(ServerLevel level, DamageSource source, float amount) { return false; }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/entity/TrainSeatEntity.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.entity

import net.minecraft.network.protocol.Packet
import net.minecraft.network.protocol.game.ClientGamePacketListener
import net.minecraft.network.protocol.game.ClientboundAddEntityPacket
import net.minecraft.network.syncher.EntityDataAccessor
import net.minecraft.network.syncher.EntityDataSerializers
import net.minecraft.network.syncher.SynchedEntityData
import net.minecraft.server.level.ServerEntity
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResult
import net.minecraft.world.damagesource.DamageSource
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.EntityDimensions
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.Pose
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.Level
import net.minecraft.world.level.storage.ValueInput
import net.minecraft.world.level.storage.ValueOutput
import net.minecraft.world.phys.Vec3

open class TrainSeatEntity(type: EntityType<out TrainSeatEntity>, level: Level) : Entity(type, level) {
    @JvmField
    var train: TrainEntity? = null
    private var cachedTrainId: Int = -1

    init {
        noPhysics = true
        setNoGravity(true)
    }

    override fun defineSynchedData(builder: SynchedEntityData.Builder) {
        builder.define(SEAT_INDEX, -1)
        builder.define(TRAIN_ID, -1)
    }

    open fun setTrain(train: TrainEntity?, seatIndex: Int) {
        this.train = train
        cachedTrainId = train?.id ?: -1
        entityData.set(TRAIN_ID, cachedTrainId)
        entityData.set(SEAT_INDEX, seatIndex)
    }

    open fun getTrain(): TrainEntity? {
        train?.let { return it }
        val vehicle = vehicle
        if (vehicle is TrainEntity) {
            train = vehicle
            return vehicle
        }
        val trainId = entityData.get(TRAIN_ID)
        if (trainId >= 0) {
            cachedTrainId = trainId
            val resolved = level().getEntity(trainId)
            if (resolved is TrainEntity) {
                train = resolved
                return resolved
            }
        }
        return null
    }

    open fun getSeatIndex(): Int = entityData.get(SEAT_INDEX)

    open fun getPassengersRidingOffset(): Double = 0.7

    override fun interact(player: Player, hand: InteractionHand, location: Vec3): InteractionResult {
        if (player.isCrouching) return InteractionResult.PASS
        val train = getTrain()
        if (train != null) {
            return train.rideSeat(player, getSeatIndex())
        }
        if (!level().isClientSide) {
            player.startRiding(this, true, false)
            return InteractionResult.CONSUME
        }
        return InteractionResult.SUCCESS
    }

    override fun tick() {
        super.tick()
        val train = getTrain()
        if (train == null || !train.isAlive || train.isRemoved) {
            if (!level().isClientSide) {
                ejectPassengers()
                discard()
            }
            return
        }
    }

    override fun readAdditionalSaveData(tag: ValueInput) {
    }

    override fun addAdditionalSaveData(tag: ValueOutput) {
    }

    override fun getAddEntityPacket(entity: ServerEntity): Packet<ClientGamePacketListener> =
        ClientboundAddEntityPacket(this, entity)

    override fun getDimensions(pose: Pose): EntityDimensions =
        EntityDimensions.scalable(HITBOX_WIDTH, HITBOX_HEIGHT)

    override fun isPushable(): Boolean = false

    override fun isPickable(): Boolean = !isRemoved

    open fun canBeCollidedWith(): Boolean = !isRemoved

    override fun canBeCollidedWith(entity: Entity?): Boolean {
        if (isRemoved) return false
        val owner = getTrain()
        if (entity === owner) return false
        if (entity != null && owner?.hasPassenger(entity) == true) return false
        return true
    }

    open fun attachToTrain(train: TrainEntity?, seatIndex: Int) {
        setTrain(train, seatIndex)
    }

    open fun belongsToTrain(trainId: Int): Boolean =
        entityData.get(TRAIN_ID) == trainId

    override fun addPassenger(passenger: Entity) {
        super.addPassenger(passenger)
        if (passenger is Player) {
            passenger.yRot = yRot
        }
    }

    override fun hurtServer(level: ServerLevel, source: DamageSource, amount: Float): Boolean {
        val train = getTrain()
        if (train != null) {
            return train.hurtServer(level, source, amount)
        }
        return false
    }

    companion object {
        private const val HITBOX_WIDTH = 0.9f
        private const val HITBOX_HEIGHT = 0.25f

        @JvmField
        val SEAT_INDEX: EntityDataAccessor<Int> =
            SynchedEntityData.defineId(TrainSeatEntity::class.java, EntityDataSerializers.INT)

        @JvmField
        val TRAIN_ID: EntityDataAccessor<Int> =
            SynchedEntityData.defineId(TrainSeatEntity::class.java, EntityDataSerializers.INT)
    }
}
